### ISSUE TITLE
I can't use pip on WSL Ubuntu

### ISSUE THREAD

**Issue by hwooo** [microsoft/WSL#4020]
**Please fill out the below information:**
* Your Windows build number:  `10.0.17763.475` / WSL Ubuntu `18.04.2`

* What you're doing and what's happening: Type `sudo pip3 install [package]`

* What's wrong / what should be happening instead:
Following error message comes out when trying to install packages via pip.

```
Collecting [package]
  Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ProtocolError('Connection aborted.', OSError(0, 'Error'))': /simple/[package]/
  Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ProtocolError('Connection aborted.', OSError(0, 'Error'))': /simple/[package]/
  Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ProtocolError('Connection aborted.', OSError(0, 'Error'))': /simple/[package]/
  Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ProtocolError('Connection aborted.', OSError(0, 'Error'))': /simple/[package]/
  Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ProtocolError('Connection aborted.', OSError(0, 'Error'))': /simple/[package]/
  Could not find a version that satisfies the requirement flask-restful (from versions: )
No matching distribution found for [package]
```

None of these solutions work.

- Re-install pip
- Re-install WSL
- sudo pip3 install [package]
- pip3 install [package] with root account
- sudo -H pip3 install [package]
- pip install --user [package]

These commands works normally.

- python -V (3.6)
- pip3 --version (9.0.1)

**Comment by Biswa96** [comment]
Allow python in Windows Firewall. Link https://docs.microsoft.com/en-us/windows/wsl/troubleshooting

**Comment by NDelt** [comment]
@Biswa96 How do I allow WSL Python to Windows Firewall? Is it possible to access WSL file from Windows?

**Comment by Biswa96** [comment]
Allow this path in Windows Firewall:

    "C:\Users\username\AppData\Local\Packages\CanonicalGroupLimited.Ubuntu18.04onWindows_79rhkp1fndgsc\LocalState\rootfs\usr\bin\python3.6"

The path may vary in Ubuntu versions from Windows Store. Explanation here: https://superuser.com/a/1325042/726810

**Comment by MatheusRV** [comment]
This could work

```
sudo apt -y purge python-pip
sudo python -m pip uninstall pip
sudo apt -y install python-pip
pip install --upgrade pip
echo "export PATH=\"${HOME}/.local/bin:$PATH\"" >>"${HOME}"/.bashrc

```

**Comment by NDelt** [comment]
@Biswa96 I can't configure the path.
![firewall](https://user-images.githubusercontent.com/46282492/57206414-782c7080-7000-11e9-9b4d-16b02b2fb2c1.PNG)

**Comment by NDelt** [comment]
@MatheusRV `pip install --upgrade pip` gives me same error message 😢

**Comment by Biswa96** [comment]
Oh... Change the `username` to your user name in your PC

**Comment by NDelt** [comment]
That's my mistake...But result is same.

**Comment by NDelt** [comment]
I've turned off Windows Defender Firewall, but nothing has changed.

**Comment by NDelt** [comment]
The cause was my Kaspersky app. If I turn off it, everything works well. I've raised this issue to Kaspersky.

**Comment by tuxdinosaur** [comment]
> This worked for me.
> `sudo apt-get update`
> `sudo apt-get upgrade`
> `sudo apt install python3-pip`

Thanks, this worked for me

**Comment by corneliusroemer** [comment]
Ok, so this issue was silent and closed for more than a year and now suddenly it garners a lot of attention again.

Simply updating and upgrading Ubuntu seems to be all that's required this time. 

In my case the error message was slightly different, the connection was broken by NewConnectionError as opposed to ProtocolError. Maybe this helps people find this issue and workarounds quicker.
```
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 
'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f81caa39bb0>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/qiskit/
```

**Comment by Bishwas-py** [comment]
Try using `sudo apt-get update && upgrade` before installation. If you're still having issue, check this forum post to [install pip in wsl](https://webmatrices.com/d/135-wsl-install-pip/2).

**Comment by theosanderson** [comment]
I was getting this with a new conda environment created with `python=3` (which gave 3.9). Using `conda create -n myEnvironment python=3.8` resolved this.

**Comment by Agrover112** [comment]
Idk why but every now and then I find it impossible to install via pip3

**Comment by iongion** [comment]
If I manually set nameservers, all works properly.

My `/etc/wsl.conf`

```ini
[network]
generateResolvConf = false
```

And the `/etc/resolv.conf`

```
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false

nameserver 1.1.1.1
nameserver 8.8.8.8
nameserver 8.8.4.4
nameserver 172.23.0.1
```

Where `1.1.1.1` is cloudflare, `8.8.8.8` and `8.8.4.4` are google, while the `172.23.0.1` is the WSL default one

**Comment by mozarik** [comment]
I have some issues with this too. My issues is pip install take too long to process (no output for 5 minute)  after some searching I find that if i disable my Public Network Firewall it seems fine again.

Does anyone know how to allow the pip (wsl) in firewall. so i can turn on my public network firewall again ?

i already try what @Biswa96. but my `*/LocalState` folder is empty. and already try what @iongion solution. but it still does not work.

**Comment by Agrover112** [comment]
> If I manually set nameservers, all works properly.
> 
> My `/etc/wsl.conf`
> 
> ```ini
> [network]
> generateResolvConf = false
> ```
> 
> And the `/etc/resolv.conf`
> 
> ```
> # This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
> # [network]
> # generateResolvConf = false
> 
> nameserver 1.1.1.1
> nameserver 8.8.8.8
> nameserver 8.8.4.4
> nameserver 172.23.0.1
> ```
> 
> Where `1.1.1.1` is cloudflare, `8.8.8.8` and `8.8.4.4` are google, while the `172.23.0.1` is the WSL default one

I'm gonna report how this has an impact on pip

**Comment by Agrover112** [comment]
> If I manually set nameservers, all works properly.
> 
> My `/etc/wsl.conf`
> 
> ```ini
> [network]
> generateResolvConf = false
> ```
> 
> And the `/etc/resolv.conf`
> 
> ```
> # This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
> # [network]
> # generateResolvConf = false
> 
> nameserver 1.1.1.1
> nameserver 8.8.8.8
> nameserver 8.8.4.4
> nameserver 172.23.0.1
> ```
> 
> Where `1.1.1.1` is cloudflare, `8.8.8.8` and `8.8.4.4` are google, while the `172.23.0.1` is the WSL default one
> 

This is indeed speeding up the PIP related issues

**Comment by mlongval** [comment]
I was having trouble with this also. 

I would run this command: 

`$ python3 -m pip --verbose install pyuic5-tool`

and nothing would happen. 

So after some reading of similar situations, I TEMPORARILY disabled the firewall on Public Networks in Windows Defender. 

![image](https://user-images.githubusercontent.com/3018682/114339332-eb12c000-9b22-11eb-9aec-e9e01b8d20f2.png)

Once the firewall was disabled I reran the pip command and everything worked as it should. 

I am unaware of a way to manually allow  WSL2 python3 through the firewall. 

I looked into changing the /etc/wsl.conf file but in my installation of WSL2 there is no such file. 

Hope this helps.

**Comment by Agrover112** [comment]
> I was having trouble with this also.
> 
> I would run this command:
> 
> `$ python3 -m pip --verbose install pyuic5-tool`
> 
> and nothing would happen.
> 
> So after some reading of similar situations, I TEMPORARILY disabled the firewall on Public Networks in Windows Defender.
> 
> ![image](https://user-images.githubusercontent.com/3018682/114339332-eb12c000-9b22-11eb-9aec-e9e01b8d20f2.png)
> 
> Once the firewall was disabled I reran the pip command and everything worked as it should.
> 
> I am unaware of a way to manually allow WSL2 python3 through the firewall.
> 
> I looked into changing the /etc/wsl.conf file but in my installation of WSL2 there is no such file.
> 
> Hope this helps.

I think it is a combination of this and @iongion  command which should work

**Comment by mobiuscreek** [comment]
The DNS solution does not work on Windows 10 with WSL2 and Ubuntu 20.04. What works is disabling the firewall which is not an ideal solution.

**Comment by pcuci** [comment]
How to achieve this in wsl2? The filesystem seems to work differently... can't enter the `ext4` volume like in wsl1

![image](https://user-images.githubusercontent.com/2169849/119360829-e122c780-bc5f-11eb-9c22-a27d0ba69712.png)

I can `pip install` when I temporarily turn off the public firewall, but I'd rather not do that. Does anyone know why `pip` depends on an incoming connection anyway?

Otherwise `pip install` simply hangs with no output, regardless of verbosity level...

**Comment by heitormagaldi** [comment]
> This worked for me.
> `sudo apt-get update`
> `sudo apt-get upgrade`
> `sudo apt install python3-pip`

me too! Thank's guy!

**Comment by OgulcanCelik** [comment]
> > I was having trouble with this also.
> > I would run this command:
> > `$ python3 -m pip --verbose install pyuic5-tool`
> > and nothing would happen.
> > So after some reading of similar situations, I TEMPORARILY disabled the firewall on Public Networks in Windows Defender.
> > ![image](https://user-images.githubusercontent.com/3018682/114339332-eb12c000-9b22-11eb-9aec-e9e01b8d20f2.png)
> > Once the firewall was disabled I reran the pip command and everything worked as it should.
> > I am unaware of a way to manually allow WSL2 python3 through the firewall.
> > I looked into changing the /etc/wsl.conf file but in my installation of WSL2 there is no such file.
> > Hope this helps.
> 
> I think it is a combination of this and @iongion command which should work

this also doesn't work on my wsl config. which is windows 10 19043. wsl2. ubuntu 20.04

I also tried apt purge python, pip, reinstalling, apt update && upgrade, nothing works.

Even with `--verbose` I don't see any output, pip just hangs in there.

**Comment by piksel** [comment]
For me it turns out that it was the keychain/dbus that prevented pip from running. It should be possible to disable it using
```
PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring pip install <PACKAGE>
```
but that didn't work for me either. I ended up uninstalling `python3-keyring`, which solved the issue:
```
sudo apt remove python3-keyring
```

**Comment by mikazuki-augus** [comment]
> For me it turns out that it was the keychain/dbus that prevented pip from running. It should be possible to disable it using
> 
> ```
> PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring pip install <PACKAGE>
> ```
> 
> but that didn't work for me either. I ended up uninstalling `python3-keyring`, which solved the issue:
> 
> ```
> sudo apt remove python3-keyring
> ```

This works for me

**Comment by bbminhtri** [comment]
> For me it turns out that it was the keychain/dbus that prevented pip from running. It should be possible to disable it using
> 
> ```
> PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring pip install <PACKAGE>
> ```
> 
> but that didn't work for me either. I ended up uninstalling `python3-keyring`, which solved the issue:
> 
> ```
> sudo apt remove python3-keyring
> ```

This works for me as well.

**Comment by rohitkrishna094** [comment]
> For me it turns out that it was the keychain/dbus that prevented pip from running. It should be possible to disable it using
> 
> ```
> PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring pip install <PACKAGE>
> ```
> 
> but that didn't work for me either. I ended up uninstalling `python3-keyring`, which solved the issue:
> 
> ```
> sudo apt remove python3-keyring
> ```
> 
> Edit: If it worked for you, just leave a reaction to the comment to avoid filling the issue with just quotes of my comment 😅

This worked for me as well.

**Comment by TOTOleHero** [comment]
Hello everybody, 

To use pip3 on WSL2 you need to force using the ".exe" binary on your command line

```
pip                 pip.exe     pip3              pip3.8            pip3.8.exe     pip3.exe
^ not work    ^ work     ^ not work   ^ not work     ^ work          ^ work 
```

**Comment by Agrover112** [comment]
> Hello everybody, 
> 
> To use pip3 on WSL2 you need to force using the ".exe" binary on your command line
> 
> ```
> pip            pip.exe     pip3         pip3.8         pip3.8.exe     pip3.exe
> ^ not work     ^ work      ^ not work   ^ not work     ^ work         ^ work
> ``` 
> 
> 

Dis you try what  @iongion  did?

**Comment by TOTOleHero** [comment]
> Dis you try what @iongion did?

I has no problème with the network ;). juste nothing when use "pip3" without ".exe" extension

**Comment by Agrover112** [comment]
> > Dis you try what @iongion did?
> 
> I has no problème with the network ;). juste nothing when use "pip3" without ".exe" extension
> 

WSL is just wierd somtimes

**Comment by amngithub** [comment]
**Solution**

_# Step 1:_
sudo apt install python3-pip

_# Step 2:_
pip3 --version

_# Step 3:_
pip3 install <packagename>
example: pip3 install notebook

**Comment by LJHG** [comment]
> I was having trouble with this also.
> 
> I would run this command:
> 
> `$ python3 -m pip --verbose install pyuic5-tool`
> 
> and nothing would happen.
> 
> So after some reading of similar situations, I TEMPORARILY disabled the firewall on Public Networks in Windows Defender.
> 
> ![image](https://user-images.githubusercontent.com/3018682/114339332-eb12c000-9b22-11eb-9aec-e9e01b8d20f2.png)
> 
> Once the firewall was disabled I reran the pip command and everything worked as it should.
> 
> I am unaware of a way to manually allow WSL2 python3 through the firewall.
> 
> I looked into changing the /etc/wsl.conf file but in my installation of WSL2 there is no such file.
> 
> Hope this helps.

this works for me

**Comment by IlyesBenighil** [comment]
> For me it turns out that it was the keychain/dbus that prevented pip from running. It should be possible to disable it using
> 
> ```
> PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring pip install <PACKAGE>
> ```
> 
> but that didn't work for me either. I ended up uninstalling `python3-keyring`, which solved the issue:
> 
> ```
> sudo apt remove python3-keyring
> ```
> 
> Edit: If it worked for you, just leave a reaction to the comment to avoid filling the issue with just quotes of my comment 😅

you are a legend

**Comment by LucaCeccagnoli** [comment]
My pip has become extremely slow at installing packages, and none of the solutions in this thread worked for me.
Here is the output for installing flask:

    Looking in indexes: https://pypi.org/simple, https://piwheels.org/simple
    Looking in links: http://pypi.pinaxproject.com, http://pypi2.pinaxproject.com
    WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.org', port=443): Read timed out. (read timeout=60.0)")': /simple/flask/
    WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.org', port=443): Read timed out. (read timeout=60.0)")': /simple/flask/
    Collecting flask
      Using cached https://piwheels.org/simple/flask/Flask-2.0.2-py3-none-any.whl (95 kB)
    Requirement already satisfied: click>=7.1.2 in ./venv/lib/python3.8/site-packages (from flask) (8.0.3)
    Requirement already satisfied: itsdangerous>=2.0 in ./venv/lib/python3.8/site-packages (from flask) (2.0.1)
    Requirement already satisfied: Werkzeug>=2.0 in ./venv/lib/python3.8/site-packages (from flask) (2.0.2)
    Requirement already satisfied: Jinja2>=3.0 in ./venv/lib/python3.8/site-packages (from flask) (3.0.3)
    Requirement already satisfied: MarkupSafe>=2.0 in ./venv/lib/python3.8/site-packages (from Jinja2>=3.0->flask) (2.0.1)
    Installing collected packages: flask
    Successfully installed flask-2.0.2

**Comment by zabbidou** [comment]
Hello, I am having the same issue. Pip hangs on every command except version.

I tried all the solutions here: I reinstalled everything, I turned off the firewall (I have the windows one turned off by default, I use Bitdefender and I also tried turning it off), I uninstalled keyring, I disabled IPv6 per a stack overflow post. Also a wild solution I found was to clear the DISPLAY variable, don't know how it's related but I tried with a functioning Xserver and a cleared variable.

My Windows is W10 Home Version 10.0.19042 Build 19042, WSL2 with Ubuntu 20.04 LTS from the Windows Store.

My pip install hangs on this:
`Starting new HTTPS connection (1): pypi.org:443`
If I Ctrl C the operation, the traceback shows that it was hanging in this function:
``` 
  File "/usr/lib/python3.8/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/usr/lib/python3.8/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/usr/lib/python3.8/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()```

**Comment by sdgMihai** [comment]
On win10, Education, ver. 20H2, build 19042.1165, I had network unreachable on pip. After checking and founding ping also returned network unreachable, I discovered that simply stopping the wireless connection and staying on Ethernet was the solution.

**Comment by LukaszPienkowski** [comment]
> For me it turns out that it was the keychain/dbus that prevented pip from running. It should be possible to disable it using
> 
> ```
> PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring pip install <PACKAGE>
> ```
> 
> but that didn't work for me either. I ended up uninstalling `python3-keyring`, which solved the issue:
> 
> ```
> sudo apt remove python3-keyring
> ```
> 
> Edit: If it worked for you, just leave a reaction to the comment to avoid filling the issue with just quotes of my comment 😅

Save me from a big headache, big thanks!

**Comment by anderl80** [comment]
I now have the same issue, it has worked for a long time but now stopped working with latest updates.

It seems in my Ubuntu WSL, the pip belonging to my Miniconda's base installation works fine, but when creating a new one `conda create -n myenv python=3` and doing `conda activate myenv` and `pip install somepippackage` it shows the error.

**Comment by tzaumiaan** [comment]
> For me it turns out that it was the keychain/dbus that prevented pip from running. It should be possible to disable it using
> 
> ```
> PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring pip install <PACKAGE>
> ```
> 
> but that didn't work for me either. I ended up uninstalling `python3-keyring`, which solved the issue:
> 
> ```
> sudo apt remove python3-keyring
> ```
> 
> Edit: If it worked for you, just leave a reaction to the comment to avoid filling the issue with just quotes of my comment 😅

Works even if i use `conda` instead of system python3. It seems the system `python3-keyring` did block something. By the way, before trying this, i have 2 conda environments with exactly the same version of python, pip, certifi packages, and one works originally while the other showed the pip networking issue. Now both work! :)

**Comment by unoebauer** [comment]
> I now have the same issue, it has worked for a long time but now stopped working with latest updates.
> 
> It seems in my Ubuntu WSL, the pip belonging to my Miniconda's base installation works fine, but when creating a new one `conda create -n myenv python=3` and doing `conda activate myenv` and `pip install somepippackage` it shows the error.

I had the exact same issue as @anderl80 and can confirm what @tzaumiaan reported: after removing the system ``python3-keyring`` (as originally found by @piksel) and restarting wsl, ``pip install`` finally worked in all ``conda`` envs, not only the base one.

**Comment by tzaumiaan** [comment]
Hi Ulrich,

If Vonda base environment has no python nor pip installed, this shouldn’t
work, right?

Ulrich Noebauer ***@***.***> schrieb am Mi. 27. Juli 2022 um
22:35:

**Comment by eneipereira** [comment]
> For me it turns out that it was the keychain/dbus that prevented pip from running. It should be possible to disable it using
> 
> ```
> PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring pip install <PACKAGE>
> ```
> 
> but that didn't work for me either. I ended up uninstalling `python3-keyring`, which solved the issue:
> 
> ```
> sudo apt remove python3-keyring
> ```
> 
> Edit: If it worked for you, just leave a reaction to the comment to avoid filling the issue with just quotes of my comment 😅

I didn't know what it really do, but it worked for me. I was trying to install flake8 but without success. After running that command i could install it.

**Comment by eneipereira** [comment]
> For me it turns out that it was the keychain/dbus that prevented pip from running. It should be possible to disable it using
> 
> ```
> PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring pip install <PACKAGE>
> ```
> 
> but that didn't work for me either. I ended up uninstalling `python3-keyring`, which solved the issue:
> 
> ```
> sudo apt remove python3-keyring
> ```
> 
> Edit: If it worked for you, just leave a reaction to the comment to avoid filling the issue with just quotes of my comment 😅

I don't know what it really do, but it worked for me. I was trying to install flake8 but without success. After running that command i've installed it.

**Comment by Anutrix** [comment]
`pip list` and `pip freeze` on a fresh venv environment takes at least 8 seconds every time,
Installing numpy(`pip install numpy`) took 55 seconds.

Clear the DISPLAY variable also didn't work.

**Comment by blackstorm0514** [comment]
```
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.org', port=443): Read timed out. (read timeout=15)")': /simple/ecdsa/
ERROR: Could not find a version that satisfies the requirement ecdsa (from versions: none)
ERROR: No matching distribution found for ecdsa
```

I tried all above solutions but still ReadTimeOut Error on my WSL.

Firewall is all off.
pip install works perfectly on my Windows terminal.
ping pypi.org also works on WSL.

I really don't understand why it acts if pypi.org is blocked on my WSL.

Anyone found another solution for this?

Thanks for any help.

**Comment by yct72** [comment]
> ```
> WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.org', port=443): Read timed out. (read timeout=15)")': /simple/ecdsa/
> ERROR: Could not find a version that satisfies the requirement ecdsa (from versions: none)
> ERROR: No matching distribution found for ecdsa
> ```
> 
> I tried all above solutions but still ReadTimeOut Error on my WSL.
> 
> Firewall is all off. pip install works perfectly on my Windows terminal. ping pypi.org also works on WSL.
> 
> I really don't understand why it acts if pypi.org is blocked on my WSL.
> 
> Anyone found another solution for this?
> 
> Thanks for any help.

Same here. Tried all above but still got error.
What I've tried:

1. The [DISPLAY](https://stackoverflow.com/questions/68016708/on-wsl2-pip-install-virtualenv-comand-hangs-for-too-long) env variable.
2. [KEYRING](https://github.com/microsoft/WSL/issues/4020#issuecomment-1100759677) described above
3. disable firewall
4. [resolv.conf](https://github.com/microsoft/WSL/issues/4020#issuecomment-809161733) described above
5. reinstall pip

environment: WSL2, win11

Thanks for any help!

**Comment by lrsperanza** [comment]
> This worked for me. `sudo apt-get update` `sudo apt-get upgrade` `sudo apt install python3-pip`

This didn't work. Had to:
- open powershell on windows
- run wsl --update

**Comment by joyehsu** [comment]
Hi there,

My case is quite similar to @zabbidou, TLS handshake just didn't work but everything other was fine.
With `curl https://pypi.org/ -vi`, client send hello but server site's hello never came back till timeout:
> *   Trying 151.101.128.223:443...
> * TCP_NODELAY set
> * Connected to pypi.org (151.101.128.223) port 443 (#0)
> * ALPN, offering h2
> * ALPN, offering http/1.1
> * successfully set certificate verify locations:
> *   CAfile: /etc/ssl/certs/ca-certificates.crt
>   CApath: /etc/ssl/certs
> * TLSv1.3 (OUT), TLS handshake, Client hello (1):
> * Operation timed out after 300164 milliseconds with 0 out of 0 bytes received
> * Closing connection 0
> curl: (28) Operation timed out after 300164 milliseconds with 0 out of 0 bytes received

After decreasing MTU size finally works for me (referred to [article](https://sbulav.github.io/wsl/wsl2-vm-tls-handshake-failed/))
here's my solution hope this can help.

Run the command under windows prompt, will show your available interfaces where can find the smallest MTU from.
`netsh interface ipv4 show subinterfaces`
for my case, it's Hinet as ADSL WAN which it's MTU is 1480
>    MTU  MediaSenseState  接收的位元組 傳送的位元組 介面
> ------  ---------------  ---------  ---------  -------------
>   1500                5          0          0  VPN - VPN Client
>   1480                1  180923235   43242607  Hinet
> 4294967295                1          0     879008  Loopback Pseudo-Interface 1
>   1500                1      36110     327707  Ethernet
>   1500                1    2072521    5569538  vEthernet (WSL)
> 

Reset MTU to eth0 with new MTU:
`sudo ip link set dev eth0 mtu 1480`

Then it's done, try pip install to see if it works.

My ENV:
- Win10
- WSL2 / ubuntu 20.04
- python 3.8 / pip 22.0.4

**Comment by olvez** [comment]
No matter how many times I restart WSL, Windows automatically changes its name server and assign by itself.

So I change little bit as Belows,
in Windows Security > Firewall & Network Protection 
In Public & private network turn off this 👇🏻 
**Block all incoming connections, including those in the list of allowed apps.**

Restart WSL and its work like charm


Env :-
Win10
WSL2 / ubuntu 22.04

**Comment by rraadd88** [comment]
Within conda environment, I get this error when I use pip for the first time. 
To fix it
1. Install a python package using conda e.g. `conda install matplotlib`
2. Then use pip install.

Env:
Win11 WSL2 / ubuntu 22.04
Firewall and proxy: off

**Comment by Twetler** [comment]
> Hello everybody,
> 
> To use pip3 on WSL2 you need to force using the ".exe" binary on your command line
> 
> ```
> pip            pip.exe     pip3         pip3.8         pip3.8.exe     pip3.exe
> ^ not work     ^ work      ^ not work   ^ not work     ^ work         ^ work
> ```
> 
> You need to install Python under windows before
> 
> ```
> me@computer:~$ whereis pip3.8
> pip3: /usr/bin/pip3 /usr/local/bin/pip3 /usr/local/bin/pip3.8 /opt/az/bin/pip3 /opt/az/bin/pip3.6 /mnt/c/Python38/Scripts/pip3.exe /usr/share/man/man1/pip3.1.gz
> me@computer:~$ whereis pip3.8.exe
> pip3.8: /usr/local/bin/pip3.8 /mnt/c/Python38/Scripts/pip3.8.exe
> ```

This has solved my  ERROR: Could not install packages due to an OSError when trying to pip install in WSL2

**Comment by sageil** [comment]
I am using Python3.10 and this worked for me.

`curl -sS https://bootstrap.pypa.io/get-pip.py | python3.10`

Chances are you will have issues with pipenv installation, to resolve those, if they occur, use `python3 -m pip install pipenv`

**Comment by dstkibbrom** [comment]
I have been having this problem and did add all my proxy settings to /etc/environment and name resolutions to /etc/resolve.conf. But that was not enough for pip to work. But exporting the https_proxy in the terminal did the magic. I am not sure why this worked though. 

'''
export https_proxy="https://proxyIP:port"
'''

**Comment by subhojeet-chowdhury** [comment]
> This worked for me. `sudo apt-get update` `sudo apt-get upgrade` `sudo apt install python3-pip`

Thankx a lot brother.

**Comment by ikitcheng** [comment]
Hi, for me, the thing that fixed it was disconnecting my VPN connection and just connecting to wifi directly.

**Comment by Abhinavammu** [comment]
I got error in upgrading pip and I can't install any pip packages. I remove it and re installed it but it saying error. Please tell me any Solutions

**Comment by craigloewen-msft** [comment]
Hey folks this issue is stale so we're closing it for our bookkeeping. It looks like there are already good answers here, if you're still seeing this please reopen a new issue for us to properly triage it! Thank you!
